src/link.ts

```diff
@@ -40,103 +40,106 @@ class Pipe<In, Out> {
     this._fn = fn;
   }
 
-  map<R>(fn: (value: Out) => R): Pipe<In, R> {
-    return new Pipe<In, R>(
+  map<T extends In, R>(fn: (value: T) => R): Pipe<T, R> {
```

Comment:
Nice idea! I'll try to adopt it. 

The `In` generic argument is an argument of first incoming value of the whole pipeline. So, unfortunately, we can't use `value: T` here. 

```pipe().map((num) = `${num}`).map(str => str)```

Here the whole pipe is `Pipe<Number, String>`, because it accepts `Number` and return `String`. But second map already accepts `String` but still returns  `Pipe<Number, String>`. I hope it makes sense :smile_cat:

Reply:
Yeah, I get it 🤔 will think about it as well!